File: harnessOS/sdk/python/harnessos_client/protocol_snapshot.py
"""SDK-local protocol surface snapshot.

The runtime SDK must not import harnessOS server internals. Tests compare this
snapshot against the server schema registry.
"""

from __future__ import annotations

DEFAULT_METHODS = {
    "session.start",
    "turn.start",
    "events.subscribe",
    "artifact.list",
    "artifact.read_metadata",
    "artifact.register_external",
    "artifact.lineage",
    "job.get",
    "job.list",
    "approval.respond",
    "connector.health",
    "pack.list",
    "pack.get",
}

FORBIDDEN_PATTERNS = (
    "meeting.",
    "knowledge.",
)

FORBIDDEN_METHODS = {
    "approval.approve",
    "approval.reject",
    "workflow.execute_stub",
    "pack.execute_stub",
    "meeting.process_recording",
    "meeting.process_audio_dir",
    "meeting.analyze_text",
    "meeting.capabilities",
}

WRAPPER_METHODS = {
    "session_start": "session.start",
    "turn_start": "turn.start",
    "events_subscribe": "events.subscribe",
    "artifact_list": "artifact.list",
    "artifact_read_metadata": "artifact.read_metadata",
    "artifact_register_external": "artifact.register_external",
    "artifact_lineage": "artifact.lineage",
    "job_get": "job.get",
    "job_list": "job.list",
    "approval_respond": "approval.respond",
    "connector_health": "connector.health",
    "pack_list": "pack.list",
    "pack_get": "pack.get",
}


def is_default_method(method: str) -> bool:
    return method in DEFAULT_METHODS


def is_forbidden_method(method: str) -> bool:
    return method in FORBIDDEN_METHODS or any(method.startswith(prefix) for prefix in FORBIDDEN_PATTERNS)

File: harnessOS/sdk/typescript/src/protocolSnapshot.ts
export const DEFAULT_METHODS = new Set([
  "session.start",
  "turn.start",
  "events.subscribe",
  "artifact.list",
  "artifact.read_metadata",
  "artifact.register_external",
  "artifact.lineage",
  "job.get",
  "job.list",
  "approval.respond",
  "connector.health",
  "pack.list",
  "pack.get",
]);

export const FORBIDDEN_PATTERNS = ["meeting.", "knowledge."];

export const FORBIDDEN_METHODS = new Set([
  "approval.approve",
  "approval.reject",
  "workflow.execute_stub",
  "pack.execute_stub",
  "meeting.process_recording",
  "meeting.process_audio_dir",
  "meeting.analyze_text",
  "meeting.capabilities",
  "method.list",
]);

export const WRAPPER_METHODS = {
  sessionStart: "session.start",
  turnStart: "turn.start",
  eventsSubscribe: "events.subscribe",
  artifactList: "artifact.list",
  artifactReadMetadata: "artifact.read_metadata",
  artifactRegisterExternal: "artifact.register_external",
  artifactLineage: "artifact.lineage",
  jobGet: "job.get",
  jobList: "job.list",
  approvalRespond: "approval.respond",
  connectorHealth: "connector.health",
  packList: "pack.list",
  packGet: "pack.get",
} as const;

export type WrapperName = keyof typeof WRAPPER_METHODS;
export type DefaultMethod = (typeof WRAPPER_METHODS)[WrapperName];

export function isDefaultMethod(method: string): boolean {
  return DEFAULT_METHODS.has(method);
}

export function isForbiddenMethod(method: string): boolean {
  return FORBIDDEN_METHODS.has(method) || FORBIDDEN_PATTERNS.some((prefix) => method.startsWith(prefix));
}
